File: frontend/src/services/authService.ts
/**
 * Authentication Service
 * Handles login, register, logout, and token management
 */
import apiClient from './apiClient';

export interface LoginCredentials {
    username: string;
    password: string;
}

export interface RegisterData {
    username: string;
    email: string;
    password: string;
    full_name?: string;
    role: 'student' | 'tpo' | 'faculty' | 'admin';
    department?: string;
}

export interface User {
    id: string;
    username: string;
    email: string;
    role: string;
    full_name?: string;
    department?: string;
    is_active: boolean;
    created_at?: string;
    last_login?: string;
}

export interface AuthResponse {
    status: string;
    access_token: string;
    refresh_token: string;
    token_type: string;
    user: User;
}

class AuthService {
    /**
     * Login user
     */
    async login(credentials: LoginCredentials): Promise<AuthResponse> {
        const response = await apiClient.post<AuthResponse>('/api/auth/login', credentials);

        // Store tokens
        localStorage.setItem('access_token', response.access_token);
        localStorage.setItem('refresh_token', response.refresh_token);
        localStorage.setItem('user', JSON.stringify(response.user));

        return response;
    }

    /**
     * Register new user
     */
    async register(data: RegisterData): Promise<any> {
        return await apiClient.post('/api/auth/register', data);
    }

    /**
     * Logout user
     */
    logout(): void {
        localStorage.removeItem('access_token');
        localStorage.removeItem('refresh_token');
        localStorage.removeItem('user');
    }

    /**
     * Get current user info from API
     */
    async getCurrentUser(): Promise<User> {
        const response = await apiClient.get<{ status: string; user: User }>('/api/auth/me');
        localStorage.setItem('user', JSON.stringify(response.user));
        return response.user;
    }

    /**
     * Get stored user from localStorage
     */
    getStoredUser(): User | null {
        const userStr = localStorage.getItem('user');
        return userStr ? JSON.parse(userStr) : null;
    }

    /**
     * Check if user is authenticated
     */
    isAuthenticated(): boolean {
        return !!localStorage.getItem('access_token');
    }

    /**
     * Change password
     */
    async changePassword(oldPassword: string, newPassword: string): Promise<any> {
        return await apiClient.post('/api/auth/change-password', {
            old_password: oldPassword,
            new_password: newPassword,
        });
    }
}

export const authService = new AuthService();
export default authService;

File: backend/services/auth_service.py
"""
Authentication Service
Handles JWT token generation, validation, and user authentication
"""
from datetime import datetime, timedelta
from typing import Optional, Dict, Any
from jose import JWTError, jwt
from passlib.context import CryptContext
from sqlalchemy.orm import Session
import os
from dotenv import load_dotenv

from models.database_models import User

load_dotenv()

# Configuration
SECRET_KEY = os.getenv("SECRET_KEY", "your_secret_key_change_in_production")
ALGORITHM = os.getenv("ALGORITHM", "HS256")
ACCESS_TOKEN_EXPIRE_MINUTES = int(os.getenv("ACCESS_TOKEN_EXPIRE_MINUTES", 15))
REFRESH_TOKEN_EXPIRE_DAYS = int(os.getenv("REFRESH_TOKEN_EXPIRE_DAYS", 7))

# Password hashing
pwd_context = CryptContext(schemes=["bcrypt"], deprecated="auto")


class AuthService:
    """Service for authentication operations"""
    
    @staticmethod
    def hash_password(password: str) -> str:
        """Hash a password using bcrypt"""
        return pwd_context.hash(password)
    
    @staticmethod
    def verify_password(plain_password: str, hashed_password: str) -> bool:
        """Verify a password against its hash"""
        return pwd_context.verify(plain_password, hashed_password)
    
    @staticmethod
    def create_access_token(data: Dict[str, Any], expires_delta: Optional[timedelta] = None) -> str:
        """
        Create JWT access token
        
        Args:
            data: Payload data (user_id, username, role)
            expires_delta: Custom expiration time
        
        Returns:
            Encoded JWT token
        """
        to_encode = data.copy()
        
        if expires_delta:
            expire = datetime.utcnow() + expires_delta
        else:
            expire = datetime.utcnow() + timedelta(minutes=ACCESS_TOKEN_EXPIRE_MINUTES)
        
        to_encode.update({
            "exp": expire,
            "iat": datetime.utcnow(),
            "type": "access"
        })
        
        encoded_jwt = jwt.encode(to_encode, SECRET_KEY, algorithm=ALGORITHM)
        return encoded_jwt
    
    @staticmethod
    def create_refresh_token(data: Dict[str, Any]) -> str:
        """
        Create JWT refresh token
        
        Args:
            data: Payload data (user_id, username)
        
        Returns:
            Encoded JWT refresh token
        """
        to_encode = data.copy()
        expire = datetime.utcnow() + timedelta(days=REFRESH_TOKEN_EXPIRE_DAYS)
        
        to_encode.update({
            "exp": expire,
            "iat": datetime.utcnow(),
            "type": "refresh"
        })
        
        encoded_jwt = jwt.encode(to_encode, SECRET_KEY, algorithm=ALGORITHM)
        return encoded_jwt
    
    @staticmethod
    def verify_token(token: str) -> Optional[Dict[str, Any]]:
        """
        Verify and decode JWT token
        
        Args:
            token: JWT token string
        
        Returns:
            Decoded token payload or None if invalid
        """
        try:
            payload = jwt.decode(token, SECRET_KEY, algorithms=[ALGORITHM])
            return payload
        except JWTError:
            return None
    
    @staticmethod
    def authenticate_user(db: Session, username: str, password: str) -> Optional[User]:
        """
        Authenticate user with username and password
        
        Args:
            db: Database session
            username: Username
            password: Plain text password
        
        Returns:
            User object if authentication successful, None otherwise
        """
        user = db.query(User).filter(User.username == username).first()
        
        if not user:
            return None
        
        if not user.is_active:
            return None
        
        if not AuthService.verify_password(password, user.password_hash):
            return None
        
        # Update last login
        user.last_login = datetime.utcnow()
        db.commit()
        
        return user
    
    @staticmethod
    def create_user(
        db: Session,
        username: str,
        email: str,
        password: str,
        role: str,
        full_name: Optional[str] = None,
        department: Optional[str] = None
    ) -> User:
        """
        Create a new user
        
        Args:
            db: Database session
            username: Unique username
            email: Email address
            password: Plain text password (will be hashed)
            role: User role (student, tpo, faculty, admin)
            full_name: Optional full name
            department: Optional department
        
        Returns:
            Created User object
        """
        # Hash password
        password_hash = AuthService.hash_password(password)
        
        # Create user
        user = User(
            username=username,
            email=email,
            password_hash=password_hash,
            role=role,
            full_name=full_name,
            department=department,
            is_active=True
        )
        
        db.add(user)
        db.commit()
        db.refresh(user)
        
        return user
    
    @staticmethod
    def get_user_by_username(db: Session, username: str) -> Optional[User]:
        """Get user by username"""
        return db.query(User).filter(User.username == username).first()
    
    @staticmethod
    def get_user_by_email(db: Session, email: str) -> Optional[User]:
        """Get user by email"""
        return db.query(User).filter(User.email == email).first()
    
    @staticmethod
    def change_password(db: Session, user_id: str, old_password: str, new_password: str) -> bool:
        """
        Change user password
        
        Args:
            db: Database session
            user_id: User UUID
            old_password: Current password
            new_password: New password
        
        Returns:
            True if successful, False otherwise
        """
        from uuid import UUID
        user = db.query(User).filter(User.id == UUID(user_id)).first()
        
        if not user:
            return False
        
        # Verify old password
        if not AuthService.verify_password(old_password, user.password_hash):
            return False
        
        # Update password
        user.password_hash = AuthService.hash_password(new_password)
        db.commit()
        
        return True
    
    @staticmethod
    def deactivate_user(db: Session, user_id: str) -> bool:
        """Deactivate a user account"""
        from uuid import UUID
        user = db.query(User).filter(User.id == UUID(user_id)).first()
        
        if not user:
            return False
        
        user.is_active = False
        db.commit()
        return True
